src/UsesPermit2.sol

```diff
@@ -0,0 +1,162 @@
+
+    bytes32 constant _OUTPUT_TYPEHASH =
+        keccak256("Output(address token,uint256 amount,address recipient,uint32 chainId)");
+
+    string constant _WITNESS_TYPESTRING = "bytes32 witness)TokenPermissions(address token,uint256 amount)";
+
+    /// @notice Thrown when a signed Output does not match the corresponding TokenPermissions.
+    error OutputMismatch();
+
+    /// @notice The Permit2 contract address.
+    address immutable permit2Contract;
+
+    constructor(address _permit2) {
+        permit2Contract = _permit2;
+    }
+
+    /// @notice Transfer a batch of tokens using permit2.
+    /// @param outputs - the Outputs for the witness field.
+    /// @param transferDetails - the TokenPermissions for the transfer, generated based on the use-case (see `_initiateTransferDetails` and `_fillTransferDetails`).
+    /// @param permit2 - the Permit2Batch information.
+    function _permitWitnessTransferFrom(
+        IOrders.Output[] memory outputs,
+        ISignatureTransfer.SignatureTransferDetails[] memory transferDetails,
+        Permit2Batch calldata permit2
+    ) internal {
+        ISignatureTransfer(permit2Contract).permitWitnessTransferFrom(
+            permit2.permit,
+            transferDetails,
+            permit2.owner,
+            _witness(outputs),
+            _OUTPUT_WITNESS_TYPESTRING,
+            permit2.signature
+        );
+    }
+
+    /// @notice Transfer tokens using permit2.
+    /// @param witness - the pre-hashed witness field.
+    /// @param permit2 - the Permit2 information.
+    function _permitWitnessTransferFrom(bytes32 witness, Permit2 calldata permit2) internal {
+        ISignatureTransfer(permit2Contract).permitWitnessTransferFrom(
+            permit2.permit,
+            _passageTransferDetails(permit2.permit.permitted),
+            permit2.owner,
+            witness,
+            _WITNESS_TYPESTRING,
+            permit2.signature
+        );
+    }
+
+    /// @notice Encode the Output array according to EIP-712 for use as a permit2 witness.
+    /// @param outputs - the Outputs to encode.
+    /// @return witness - the encoded witness field.
+    function _witness(IOrders.Output[] memory outputs) internal pure returns (bytes32 witness) {
+        uint256 num = outputs.length;
+        bytes32[] memory hashes = new bytes32[](num);
+        for (uint256 i = 0; i < num; ++i) {
+            hashes[i] = keccak256(abi.encode(_OUTPUT_TYPEHASH, outputs[i]));
+        }
+        witness = keccak256(abi.encodePacked(hashes));
+    }
+
+    /// @notice Encode & hash the rollupChainId and rollupRecipient for use as a permit2 witness.
+    /// @return witness - the encoded witness field.
+    function _witness(uint256 rollupChainId, address rollupRecipient) internal pure returns (bytes32 witness) {
+        witness = keccak256(abi.encode(rollupChainId, rollupRecipient));
```

Comment:
chatting with Sara rn about EIP-712 encoding two "atomic types" into a single bytes32 witness hash - right now, simply encoding it in an opaque fashion

Reply:
https://github.com/init4tech/zenith/pull/63/commits/e88165b9de8a1b4c004c28ba2b45c69087d366cd